### ISSUE TITLE
Vm auth: if the request does not fully cache the body to buf due to server connection disconnection, it will not be retried.

### ISSUE THREAD

**Issue by shichanglin5** [VictoriaMetrics/VictoriaMetrics#6445]
### Describe the bug

**description in comment**:
 ```go
func (rtb *readTrackingBody) canRetry() bool {
	if rtb.cannotRetry {
		return false
	}
 // if the request does not fully cache the body to buf 
 // due to server connection disconnection, it will not be retried.
	if len(rtb.buf) > 0 && !rtb.needReadBuf {
		return false
	}
	return true
}
```

**log**
```log
2024-06-06T20:30:37+08:00 2024-06-06T12:30:37.396Z	warn	VictoriaMetrics/app/vmauth/main.go:268	remoteAddr: "172.26.13.17:64104, X-Forwarded-For: 10.214.137.216"; requestURI: /api/v1/push?job=skyeye-skynet; cannot proxy the request to http://10-205-89-19.vminsert-skyeye.vm-cluster-skyeye.svc.cluster.local.:8480/insert/2:6/prometheus/api/v1/push?job=skyeye-skynet: readfrom tcp 10.205.89.16:42960->10.205.89.19:8480: write tcp 10.205.89.16:42960->10.205.89.19:8480: write: broken pipe
```

### To Reproduce

1. do request with partial read body to buf
2. close client-server connection
3. log `cannot proxy the request to xxx`

### Version

I built it with source code：
git revision: 1a1ee87b46d4e9c76d74a1c7ddce5bb272efe5cc

### Logs

```
2024-06-06T20:30:37+08:00 2024-06-06T12:30:37.396Z	warn	VictoriaMetrics/app/vmauth/main.go:268	remoteAddr: "172.26.13.17:64104, X-Forwarded-For: 10.214.137.216"; requestURI: /api/v1/push?job=skyeye-skynet; cannot proxy the request to http://10-205-89-19.vminsert-skyeye.vm-cluster-skyeye.svc.cluster.local.:8480/insert/2:6/prometheus/api/v1/push?job=skyeye-skynet: readfrom tcp 10.205.89.16:42960->10.205.89.19:8480: write tcp 10.205.89.16:42960->10.205.89.19:8480: write: broken pipe
```

**Comment by shichanglin5** [comment]
In addition, I have observed that vm auth monitoring, go gc accounts for a relatively high proportion( 50%), and perhaps gc can be reduced by reusing buf to avoid frequent memory allocation.
<img width="752" alt="image" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/48707349/2c598aad-6945-4d43-b34b-9e86f0c3e399">

**Comment by hagen1778** [comment]
https://github.com/VictoriaMetrics/VictoriaMetrics/pull/6533 has been merged and will be included into the next release.

**Comment by shichanglin5** [comment]
good!

**Comment by f41gh7** [comment]
We keep issues  open until actual release. So re-opening it.

**Comment by valyala** [comment]
@shichanglin5 , could you build `vmauth` from the commit 7ee57974935a662896f2de40fdf613156630617d according to [these docs](https://docs.victoriametrics.com/vmauth/#how-to-build-from-sources) and verify whether it works as expected on your workload?

**Comment by shichanglin5** [comment]
> @shichanglin5 , could you build `vmauth` from the commit [7ee5797](https://github.com/VictoriaMetrics/VictoriaMetrics/commit/7ee57974935a662896f2de40fdf613156630617d) according to [these docs](https://docs.victoriametrics.com/vmauth/#how-to-build-from-sources) and verify whether it works as expected on your workload?

ok, i'll try it later

**Comment by valyala** [comment]
The commit https://github.com/VictoriaMetrics/VictoriaMetrics/commit/7ee57974935a662896f2de40fdf613156630617d has been included in [vmauth v1.102.0](https://docs.victoriametrics.com/changelog/). Closing the issue as fixed.

@shichanglin5 , please re-open the issue if it is still reproducible with [vmauth](https://docs.victoriametrics.com/vmauth/) v1.102.0.
